components/scream/src/share/tests/debug_tools_tests.cpp

```diff
@@ -0,0 +1,132 @@
+#include <catch2/catch.hpp>
+
+#include "share/scream_assert.hpp"
+#include <csignal>
+#include <unistd.h>
+#include  <csetjmp>
+#include <iostream>
+
+namespace {
+
+jmp_buf JumpBuffer;
+
+static volatile std::sig_atomic_t gSignalStatus = 0;
+
+void signal_handler (int /* signum */) {
+  gSignalStatus = 1;
+  std::longjmp(JumpBuffer,gSignalStatus);
+}
+
+void run_fpe_tests () {
+  int mask = scream::get_enabled_fpes();
+
+  std::cout << "tests mask: " << mask << "\n";
+  // Get the current fenv.
+  // Note: for some reason, each time SIGFPE is thrown or getenv
+  //       is called, the fenv is reset to 0. So remember to
+  //       set the fenv before each test.
+  std::fenv_t fenv;
+  feholdexcept(&fenv);
+  fesetenv(&fenv);
+
+  const auto has_fe_divbyzero = [](int mask)->bool{
+    return mask & FE_DIVBYZERO;
+  };
+  const auto has_fe_overflow = [](int mask)->bool{
+    return mask & FE_OVERFLOW;
+  };
+  const auto has_fe_invalid = [](int mask)->bool{
+    return mask & FE_INVALID;
+  };
+
+  std::cout << " has FE_DIVBYZERO: " << (has_fe_divbyzero(mask) ? "yes" : "no") << "\n";
+  std::cout << " has FE_INVALID: " << (has_fe_invalid(mask) ? "yes" : "no") << "\n";
+  std::cout << " has FE_OVERFLOW: " << (has_fe_overflow(mask) ? "yes" : "no") << "\n";
+
+  double one = 1.0;
+  double zero = 0.0;
+  double inf, nan, ovfl;
+
+  // Run the tests.
+  // Note: sometimes a FPE is not thrown when the bad number
+  //       is generated, but rather the next time is used.
+  //       Therefore, sometimes we multiply the result
+  //       by 1.0 before testing that the FPE was thrown.
+  
+  // Test 1/0
+  setjmp(JumpBuffer);
```

Comment:
I'm just skimming through this so I'm probably missing something, but I think this continuation will put you back at line 57 when line 58 FPEs, so you'll end up repeating line 58. My only though is that somehow after the first FPE another one won't except, so line 58 runs w/o excepting. C-style continuations are usually done like this:
```C
  if (setjmp(JumpBuffer)) {
    // Continuing.
    REQUIRE (gSignalStatus==(has_fe_divbyzero(mask) ? 1 : 0));
  ￼  gSignalStatus = 0;
￼    fesetenv(&fenv);
  } else {
    // Trigger the continuation.
    inf = one/zero;
  }
```

Reply:
I spent some time learning the signal handling procedures, and I admit I was confused as to how does the longjmp manages to go _past_ the line that raised the exception. My thought was that longjmp goes 1 line past the one where the signal was generated.

I ran the code, and it doesn't not stall, as it is. However, I did have to reset the env after each throw, or else the next time I would not see an FPE thrown...

I found a bug report in glibc, that fegetenv also resets the env (so you need to fesetenv right after a fegetenv, just to leave things as you found them), but that's separate. And I thought it should have been fixed by now (running glibc 2.17, and I think that bug was addressed in 2.2).

Reply:
Ah! I tried the way you suggested (if-else), and it works. Not only that, I no longer need to reset the fenv after each test. Thanks!